tests/digraph/test_collect_bicolor_runs.py

```diff
@@ -0,0 +1,313 @@
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License. You may obtain
+# a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+
+import unittest
+
+import retworkx
+
+
+class TestCollectBicolorRuns(unittest.TestCase):
+    def test_cycle(self):
+        dag = retworkx.PyDiGraph()
+        dag.extend_from_edge_list([(0, 1), (1, 2), (2, 0)])
+        with self.assertRaises(retworkx.DAGHasCycle):
+            retworkx.collect_bicolor_runs(dag, lambda _: True, lambda _: -1)
+
+    def test_filter_function_inner_exception(self):
+        dag = retworkx.PyDiGraph()
+        dag.add_node("a")
+        dag.add_child(0, "b", None)
+        dag.add_child(1, "c", None)
+
+        def fail_function(node):
+            raise IndexError("Things fail from time to time")
+
+        with self.assertRaises(IndexError):
+            retworkx.collect_bicolor_runs(dag, fail_function, lambda _: -1)
+
+        with self.assertRaises(IndexError):
+            retworkx.collect_bicolor_runs(dag, lambda _: True, fail_function)
+
+    def test_empty(self):
+        dag = retworkx.PyDAG()
+        self.assertEqual(
+            [], retworkx.collect_bicolor_runs(dag, lambda _: True, lambda _: -1)
+        )
+
+    def test_two_colors(self):
+        """
+        Input:
+        ┌─────────────┐                 ┌─────────────┐
+        │             │                 │             │
+        │    q0       │                 │    q1       │
+        │             │                 │             │
+        └───┬─────────┘                 └──────┬──────┘
+            │          ┌─────────────┐         │
+        q0  │          │             │         │  q1
+            │          │             │         │
+            └─────────►│     cx      │◄────────┘
+            ┌──────────┤             ├─────────┐
+            │          │             │         │
+        q0  │          └─────────────┘         │  q1
+            │                                  │
+        ┌───▼─────────┐                 ┌──────▼──────┐
+        │             │                 │             │
+        │    q0       │                 │    q1       │
+        │             │                 │             │
+        └───┬─────────┘                 └──────┬──────┘
```

Comment:
It feels strange that you included this "intermediate" qubit nodes since they don't exist in a `DAGCircuit`

Reply:
I forgot `DAGCircuit`, fortunately the code works on both test cases 😃. Updated with more accurate test cases without intermediate nodes